File: 2023/01/solution.js
module.exports = {
    solutions: [142, 281],
    one: (input) => {
        const lines = input.split("\n");
        const digits = lines.map((line) => line.match(/\d/g));
        const numbers = digits.map((arr) => {
            if (Array.isArray(arr) && arr.length) {
                return `${arr[0]}${arr.slice(-1)}`;
            }
            return 0;
        });
        return numbers.reduce((acc, cur) => acc + Number(cur), 0);
    },
    two: (input) => {
        const lines = input.split("\n").filter((line) => line.length);
        // const values = lines.map((line) => line.match(/one|two|three|four|five|six|seven|eight|nine|\d/g));
        const values = lines.map((line) => {
            const pattern = /one|two|three|four|five|six|seven|eight|nine|\d/g;
            let matches = [];
            while ((found = pattern.exec(line))) {
                matches.push(found[0]);
                pattern.lastIndex = found.index + 1;
            }
            return matches;
        });

        const digits = values.map((arr) => {
            return arr.map((val) => {
                if (val === "one") return 1;
                if (val === "two") return 2;
                if (val === "three") return 3;
                if (val === "four") return 4;
                if (val === "five") return 5;
                if (val === "six") return 6;
                if (val === "seven") return 7;
                if (val === "eight") return 8;
                if (val === "nine") return 9;
                return Number(val);
            });
        });

        const numbers = digits.map((arr) => {
            if (Array.isArray(arr) && arr.length) {
                return `${arr[0]}${arr.slice(-1)}`;
            }
            return 0;
        });

        return numbers.reduce((acc, cur) => acc + Number(cur), 0);
    },
};

File: 2023/01/solution.go
package year2023

import (
	"regexp"
	"strconv"

	"advent_of_code/utils"
)

var testSolutionPart1 int = 142
var testSolutionPart2 int = 281

func Day1() utils.Solution {
	input := utils.GetInputContent(2023, 1)
	test1 := utils.GetTestContent(2023, 1, 1)
	test2 := utils.GetTestContent(2023, 1, 2)

	return utils.Solution{
		Day:   1,
		Part1: strconv.Itoa(part1(input)),
		Part2: strconv.Itoa(part2(input)),
		Test1: part1(test1) == testSolutionPart1,
		Test2: part2(test2) == testSolutionPart2,
	}
}

func part1(input string) int {
	lines := utils.GetLines(input)

	var numbers []string

	for _, line := range lines {
		match := regexp.MustCompile(`\d`).FindAllString(line, -1)
		var first = match[0]
		var last = match[len(match)-1]
		numbers = append(numbers, first+last)
	}

	return sumSliceOfDigitStrings(numbers)
}

func part2(input string) int {
	lines := utils.GetLines(input)

	var numbers []string

	for _, line := range lines {
		r := regexp.MustCompile(`^(one|two|three|four|five|six|seven|eight|nine|\d)`)
		var lineNumbers []string
		for len(line) > 0 {
			match := r.FindString(line)
			line = line[1:]
			if match != "" {
				lineNumbers = append(lineNumbers, findDigit(match))
			}
		}
		if len(lineNumbers) != 0 {
			var first = lineNumbers[0]
			var last = lineNumbers[len(lineNumbers)-1]
			numbers = append(numbers, first+last)
		}
	}

	return sumSliceOfDigitStrings(numbers)
}

// sumSliceOfDigitStrings sums the digits in a slice of strings, assuming each string represents a Number digit
func sumSliceOfDigitStrings(digits []string) int {
	var sum int
	for _, digit := range digits {
		i, _ := strconv.Atoi(digit)
		sum = sum + i
	}
	return sum
}

// findDigit returns the digit if the string is a number word, else return the string
func findDigit(x string) string {
	switch x {
	case "one":
		return "1"
	case "two":
		return "2"
	case "three":
		return "3"
	case "four":
		return "4"
	case "five":
		return "5"
	case "six":
		return "6"
	case "seven":
		return "7"
	case "eight":
		return "8"
	case "nine":
		return "9"
	default:
		return x
	}
}
